contracts/transfer/src/transfer.rs

```diff
@@ -0,0 +1,132 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use core::convert::TryFrom;
+
+use alloc::vec::Vec;
+use canonical::{Canon, Store};
+use canonical_derive::Canon;
+use dusk_bls12_381::BlsScalar;
+use dusk_bytes::Serializable;
+use dusk_jubjub::JubJubAffine;
+use dusk_kelvin_map::Map;
+use dusk_pki::PublicKey;
+use phoenix_core::{Crossover, Fee, Message, Note};
+
+mod tree;
+use tree::Tree;
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Canon)]
+pub struct PublicKeyBytes([u8; PublicKey::SIZE]);
+
+impl From<PublicKey> for PublicKeyBytes {
+    fn from(pk: PublicKey) -> Self {
+        Self(pk.to_bytes())
+    }
+}
+
+#[derive(Debug, Default, Clone, Canon)]
+pub struct Transfer<S: Store> {
+    pub(crate) notes: Tree<S>,
+    pub(crate) notes_mapping: Map<u64, Vec<Note>, S>,
+    pub(crate) nullifiers: Map<BlsScalar, (), S>,
+    pub(crate) roots: Map<BlsScalar, (), S>,
+    pub(crate) balance: Map<BlsScalar, u64, S>,
+    pub(crate) message_mapping:
+        Map<BlsScalar, Map<PublicKeyBytes, Message, S>, S>,
+    pub(crate) message_mapping_set:
+        Map<BlsScalar, (PublicKey, JubJubAffine), S>,
+}
+
+impl<S: Store> Transfer<S> {
+    pub(crate) fn update_root(&mut self) -> Result<(), S::Error> {
```

Comment:
Why this function even exist? It does not appear in the specs and the root of the tree should only be updated by the vm once the changes to the state are applied.

Reply:
The `Store` update and `root` mappings update after execute are different routines.

This function exists to keep the root of the tree updated

Reply:
But aren't you assuming here that nodes will order the tx's on the same way? Idk wether that;s an assumption that can be done now whether it affects this TBH.

Reply:
I don't see any potential problem out of this.

Internally, I have a set of old roots. I need to update this set whenever a successful tx is executed and mutates the notes tree